### ISSUE TITLE
Dyn Dyf Convergence

### ISSUE THREAD

**Issue by ikeough** [DynamoDS/Dynamo#7626]
When I started Dynamo, there was only one file format. It was xml, and I called it .dyn. Then @Steell  came on board, and implemented the ability for a graph to be embedded in another graph as a function call. Under the covers, the main graph is an object of the type `HomeWorkspaceModel`, while a function definition is an object of the type `CustomNodeWorkspaceModel`. The differences between the two types of graph, in code, are as follows:
- Dynamo can only have one `HomeWorkspaceModel` open at a time. You can open as many `CustomNodeWorkspaceModel`s as you'd like.
- `CustomNodeWorkspaceModel` has `Input` and `Output` nodes. `Input` nodes allow you to specify the input type and optionally, a default value described as a snippet of Design Script.
- `HomeWorkspaceModel` objects serialize to .dyn files.
- `CustomNodeWorkspaceModel` objects serialize to .dyf files.

More to come...

**Comment by jnealb** [comment]
@ikeough Once upon a time, in a galaxy far, far away ... looking forward to the sequel :+1:

**Comment by Dewb** [comment]
This is likely already implicit, but it's worth calling out: the requirements around `Input` and `Output` would ideally begin to align with the Computation Graph Service's notion of workspace inputs and outputs. Specifically, the idea of typed inputs and outputs. Typeless I/O nodes may work for custom nodes, but in other contexts, some user-decoration of the expected format of the data is necessary for many of the experiences we want to build.

**Comment by andydandy74** [comment]
Oh wow. Great idea, great potential!

**Comment by ikeough** [comment]
@Dewb agreed. @gregmarr has been bugging me for a while to get the typed input and output nodes into Dynamo. The question is how to nicely dovetail those with the existing input and output nodes which take snippets of design script to define the type expectations? If we have a `BooleanInput` node, do we no longer support an `Input` node where you can say `bool : true`?

**Comment by gregmarr** [comment]
I think we decided that there would still be the concept of "user input nodes" (`BooleanInputNode`, ...) and "user output nodes" (`BooleanOutputNode`, ...) as separate from "graph input nodes" (`InputNode`) and "graph output nodes" (`OutputNode`), where we would recommend using the "user input/output nodes" where possible, but you could use the "graph input/output node" to pass more complicated data structures between graphs.

**Comment by gregmarr** [comment]
Clarification to @ikeough's comment, we already have the typed input nodes, it's only the typed output nodes that we need to add.

**Comment by aparajit-pratap** [comment]
@ikeough I understand that one of the issues with custom workspaces currently is debugging, namely, displaying warnings on nodes and nested custom nodes and bubbling them up to the home workspace. I can't remember if this had to do with custom workspaces being function definitions. I think we need to keep in mind that if we do the same for home workspaces, i.e. compiling them into function definitions, we do not break node warnings and likewise will be solving the issue of node warnings for the nested/chained workspaces (functions defs)?

**Comment by sharadkjaiswal** [comment]
By converging DYN and DYF, I think the first thing we want to achieve is to use DYN as a custom node inside another workspace. 
- I did a quick experiment by adding an ID attribute to Workspace element in DYN file and suddenly this file became a custom node and I can insert that in another workspace and run. The right most leaf nodes became the output of the custom node but there was no input. 
- Right now customizer treats sliders or few other input node types as input to the graph if it has the attribute `isSelectedInput=True`. We can update the existing custom node compilation code to treat the nodes with attribute `isSelectedInput=True` as input to the graph and then we are done with a quick prototype.

Here are few things that we would like to consider:
- It's good to treat the left most nodes as input and right most nodes as output to of the graph. However, the current custom nodes do allow you to accept any arbitrary object type as input which can't be used in this scheme of things. Maybe we can consider annotating some nodes other than extreme left node as input by adding some more data to a node such as 
`<InputSymbol value="surf: Surface" PortId = "1" Description="Input Surface for panels"/>` The portId here is to tell which output port of the node should be considered as input and `surf` is the variable name of type `Surface`.
- The default name for the DYN file is Home and hence the custom node is shown as Home, we might consider using its file name as node name or saving the file name as the default name of the workspace. We can also provide options to add a description of the file.
- I think we are still considering to have only one active DYN file, so when the referenced DYN is opened it will still be opened as a custom node and won't run by itself. 
- The File Open dialog should provide an option whether to open the DYN file as the main workspace aka `Home Workspace` or a `Custom Node Workspace`.

**Comment by ikeough** [comment]
@sharadkjaiswal Sorry it has taken me so long to see that you've commented here. 

- Can't we just require the use of `Input` and `Output` nodes and do away with the "magic" of figuring out the inputs and outputs? An `Input` or `Output` node used in the "root" graph would just be ignored. But when that graph was embedded, they would be used in defining the function.

- I agree with the idea to use the file name as the graph name. Although, we could also do away with `Home` altogether and just keep any newly opened graph in an unnamed, unsaved state. When the user saves the graph, the name of the graph is saved.

- See #7595. I think we should allow multiple open graphs. Each with their own `EngineController`.

- I think `HomeWorkspaceModel` and `CustomNodeWorkspaceModel` should be collapsed into one class.
